Quicksave/quickload feature

kajott opened the pull request:
This PR implements quicksaving and quickloading from a running game, and fixes lots of crashes due to uncached graphics chunks. The latter is required because during playtesting with quicksave/quickload, I hit these situations quite often. It's possible that the new feature exacerbates the graphics chunk cache issues, but I doubt that it causes them. I don't really understand why these things happen though, and it's certainly best on the long run to find the root cause, but so far the band-aid solution seems to work just fine.

Missing features so far:
- Quicksaving always overwrites the sixth savegame slot, not asking for confirmation if that slot isn't already a quicksave slot.
- The key bindings are not configurable yet (fixed to F5 + F9, which is what [Keengine](http://www.shikadi.net/keenwiki/KEENGINE) uses). It would be easy to make them configurable, but without a way to save the preferences, that doesn't really make sense.

sulix commented:
Thanks for this!

There shouldn't be that many uncached chunk issues. I'm seeing some stuttering when the uncached chunks are loaded, too, so I think this is something we'd want to fix. It seems to be happening because the CK_LoadGame() function doesn't actually cache anything.

This is due to an assumption that loading of games will only happen while the menu is displayed, so instead of actually caching anything during the load, the chunks are marked to be cached when the menu is exited.
(This is what the playing around with ca_levelbit and ca_levelnum is: keen keeps a stack of "cache levels", the topmost of which here will be the menu, under which is the current level. By tempotarily lowering the cache level during map loading, any chunks required are set to be loaded when the top level is popped, i.e. the menu is closed.)

For QuickLoad, we'd want to have CK_LoadGame just call CK_LoadLevel(true) to make sure everything is cached immediately. We'd want to preserve the current behaviour for loads from the menu, though, so it may be worth introducing another argument to CK_LoadGame() to separate those cases.

I think we can keep the improved code for handling uncached chunks, though, so I don't think that effort has gone to waste.

kajott commented:
If I understand you correctly, a call to `CK_LoadLevel(true)` should be sufficient to make it work without the cache issues? I'm more than happy to try this :) But does it really need to be right in `CK_LoadGame`? Shouldn't it be OK too to do it after that, but before the next iteration of the game loop, e.g. in `ck_play.c:1159`?

sulix commented:
I think it makes more sense to have it in CK_LoadGame(), as it'd need to replace the CK_LoadLevel(false) call (and the mucking around with ca_levelbit/ca_levelnum).

Basically, for the loading-from-menu case, we want to lower the cache level (with the changes to ca_levelbit, etc), and then load the level but not cache anything. For the loading-in-game case (i.e., quickload) we'd want to just call CK_LoadLevel(true), and skip the changes to ca_levelbit/ca_levelnum.

(The other option is to just skip the ca_levelbit/ca_levelnum changes, and then call CA_CacheMarks() separately, but I think it makes more sense to use CK_LoadLevel with 'true', which handles that for you. You might need to change something in CK_LoadLevel() if you want to change the loading screen, though.)

Does that make sense?

kajott commented:
Yes, that makes sense. A new parameter in `CK_LoadGame` it is, then :+1:

kajott commented:
Using `CK_LoadLevel(true)` seems to work indeed; I now only get "uncached Tile16(m)" messages when activating doors etc., which is usual even when not using any quickload stuff. I didn't do any full playthroughs to test it this time though, just a few levels.

I fixed some related issues and added configurable key bindings too.

Finally, there's a little Makefile improvement for Win32, and a rudimentary (but useful) VS Code setup. Both are not related to the feature; I can put them into their own PR if desired.

sulix commented on src/ck_play.c:
```diff
@@ -1054,7 +1054,8 @@ void CK_CheckKeys()
 
 	// Drop down status
-	if (IN_GetKeyState(IN_SC_Enter))
+	// TODO: why does this use IN_GetKeyState() instead of IN_GetLastScan()?
```
FYI, this is just because that's what the original game used (it checks its 'Keyboard' array, which is our equivalent of IN_GetKeyState(), rather than checking its 'LastScan' variable). This works properly if there are muliple keys pressed, so maybe that's why(?)

sulix commented:
Thanks!

One remaning issue: the 'key_button_controls' variable can't be found: this seems to be because it's declared const (which implies static linkage within the id_in.c file):
```
/usr/bin/ld: /home/david/Development/omnispeak/src/ck_us_2.c:686: undefined reference to `key_direction_controls'
```

Could we perhaps declare it in id_in.h (because, if it's declared extern first in something included in id_in.c, it should have the proper linkage), and possible rename it in_button_controls?

Otherwise, it'd be nice to perhaps have a comment pointing out that the status, quicksave, and quickload key bindings are new in IN_KeyMapping, so we know what ones need saving separately.

I'm happy to accept the Makefile and VSCode changes as well. I installed VSCode to give it a try, and I think I got everything working.
VSCode did spit out a bunch of warnings about the tasks.json file, though — it may be worth double checking that it's okay.

kajott commented:
This round of commits should resolve the remaining issues. (At least I hope so -- sorry for missing the build-breaking bug yesterday ...)

sulix commented on src/ck_game.c:
```diff
@@ -836,6 +836,13 @@ void CK_TryAgainMenu()
+			}
 		} /* while */
 	}
 }
```
Need a "return" statement here.

kajott replied:
Argh, of course! :/
Maybe the Makefile needs a `-Wall` by default so stuff like this is harder to miss ...

sulix commented:
Thanks!

I ended up fixing the return statement and merging this myself.
